quinn/src/streams.rs

```diff
@@ -149,6 +151,39 @@ where
         conn.wake();
     }
 
+    /// Completes if/when the peer stops the stream, yielding the error code
+    pub fn stopped(&mut self) -> Stopped<'_, S> {
+        Stopped { stream: self }
+    }
+
+    #[doc(hidden)]
+    pub fn poll_stopped(&mut self, cx: &mut Context) -> Poll<Result<VarInt, WriteError>> {
+        let mut conn = self.conn.lock().unwrap();
+
+        if self.stopped_event.is_none() {
+            match conn.inner.stopped(self.stream) {
+                Ok(Some(error_code)) => return Poll::Ready(Ok(error_code)),
+                Err(_) => return Poll::Ready(Err(WriteError::UnknownStream)),
+                _ => (),
+            }
+
+            let (send, recv) = oneshot::channel();
```

Comment:
Since the error code is now persisted in the inner `Connection`, we could simplify this a bit by replacing the oneshot table with a `Waker` table and just calling  `conn.inner.stopped(self.stream)` each poll rather than only once.